marlowe-runtime/src/Language/Marlowe/Runtime/History/Follower.hs

```diff
@@ -0,0 +1,559 @@
+data SomeContractChangesTVar = forall v. SomeContractChangesTVar (ContractChangesTVar v)
+
+mkFollower :: FollowerDependencies -> STM Follower
+mkFollower deps@FollowerDependencies{..} = do
+  someChangesVar <- newTVar Nothing
+  let
+    stInit = SendMsgRequestHandshake schemaVersion1_0 handshake
+    handshake = ClientStHandshake
+      { recvMsgHandshakeRejected = \_ -> pure $ Left HansdshakeFailed
+      , recvMsgHandshakeConfirmed = findContract
+      }
+
+    findContract = do
+      let move = FindTx $ txId $ unContractId contractId
+      pure $ SendMsgQueryNext move handleContract (pure handleContract)
+
+    handleContract = ClientStNext
+      { recvMsgQueryRejected = \err _ -> failWith $ FindTxFailed err
+      , recvMsgRollForward = \tx point _ -> case point of
+          Genesis -> error "transaction detected at Genesis"
+          At blockHeader -> case exctractCreation deps tx of
+            Left err ->
+              failWith $ ExtractContractFailed err
+            Right (SomeCreateStep version create@CreateStep{..}) -> do
+              changesVar <- atomically do
+                changesVar <- newTVar mempty { steps = Map.singleton blockHeader [Create create] }
+                writeTVar someChangesVar
+                  $ Just
+                  $ SomeContractChangesTVar
+                  $ ContractChangesTVar version changesVar
+                pure changesVar
+              let payouts = mempty
+              let scriptOutput = TransactionScriptOutput (unContractId contractId) datum
+              let previousState = Nothing
+              followContract blockHeader FollowerContext{..} FollowerState{..}
+      , recvMsgRollBackward = \_ _ -> error "Rolled back from genesis"
+      }
+
+  pure Follower
+    { runFollower = connectToChainSeek $ ChainSeekClient $ pure stInit
+    , changes = do
+        mChangesVar <- readTVar someChangesVar
+        for mChangesVar \(SomeContractChangesTVar (ContractChangesTVar version changesVar)) -> do
+          changes <- readTVar changesVar
+          writeTVar changesVar mempty
+          pure $ SomeContractChanges version changes
+    }
+
+data FollowerContext v = FollowerContext
+  { version             :: MarloweVersion v
+  , create              :: CreateStep v
+  , contractId          :: ContractId
+  , changesVar          :: TVar (ContractChanges v)
+  , scriptAddress       :: Chain.Address
+  , payoutValidatorHash :: ScriptHash
+  , slotConfig          :: SlotConfig
+  , securityParameter   :: Int
+  }
+
+data PreviousState a
+  = Retained BlockHeader a
+  | Truncated
+  deriving (Functor)
+
+data FollowerState v = FollowerState
+  { payouts       :: Map Chain.TxOutRef (Payout v)
+  , scriptOutput  :: TransactionScriptOutput v
+  , previousState :: Maybe (PreviousState (FollowerState v))
+  }
+
+data ClosedPreviousState v
+  = ClosedPreviousOpen (FollowerState v)
+  | ClosedPreviousClosed (FollowerStateClosed v)
+
+data FollowerStateClosed v = FollowerStateClosed
```

Comment:
Would it be worthwhile to move all the type declarations to a module `Language.Marlowe.Runtime.History.Follower.Types`?

Reply:
Not these ones, they are internal implementation details, not part of the API (I don't want them used elsewhere as they may change)

Reply:
In a subsequent PR, I will extract these types as they will begin to be shared some more.